src/sui/SuiTools.ts

```diff
@@ -0,0 +1,194 @@
+import { getFullnodeUrl, SuiClient, SuiHTTPTransport } from '@mysten/sui/client'
+import { Ed25519Keypair } from '@mysten/sui/keypairs/ed25519'
+import { isValidSuiAddress, parseStructTag } from '@mysten/sui/utils'
+import { div } from 'biggystring'
+import { entropyToMnemonic, validateMnemonic } from 'bip39'
+import {
+  EdgeCurrencyInfo,
+  EdgeCurrencyTools,
+  EdgeEncodeUri,
+  EdgeIo,
+  EdgeMetaToken,
+  EdgeParsedUri,
+  EdgeToken,
+  EdgeTokenMap,
+  EdgeWalletInfo,
+  JsonObject
+} from 'edge-core-js/types'
+import { base64 } from 'rfc4648'
+
+import { PluginEnvironment } from '../common/innerPlugin'
+import { asMaybeContractLocation, validateToken } from '../common/tokenHelpers'
+import { asSafeCommonWalletInfo } from '../common/types'
+import { encodeUriCommon, parseUriCommon } from '../common/uriHelpers'
+import { getLegacyDenomination, mergeDeeply } from '../common/utils'
+import { asSuiPrivateKeys, SuiInfoPayload, SuiNetworkInfo } from './suiTypes'
+
+export class SuiTools implements EdgeCurrencyTools {
+  io: EdgeIo
+  builtinTokens: EdgeTokenMap
+  currencyInfo: EdgeCurrencyInfo
+  networkInfo: SuiNetworkInfo
+  initOptions: JsonObject
+
+  suiClient: SuiClient
+
+  constructor(env: PluginEnvironment<SuiNetworkInfo>) {
+    const { builtinTokens, currencyInfo, initOptions, io, networkInfo } = env
+    this.io = io
+    this.currencyInfo = currencyInfo
+    this.builtinTokens = builtinTokens
+    this.networkInfo = networkInfo
+    this.initOptions = initOptions
+
+    const suiTransport = new SuiHTTPTransport({
+      url: getFullnodeUrl(networkInfo.network),
+      fetch: io.fetch as typeof fetch
+    })
+    this.suiClient = new SuiClient({
+      transport: suiTransport
+    })
+  }
+
+  async getDisplayPrivateKey(
+    privateWalletInfo: EdgeWalletInfo
+  ): Promise<string> {
+    const { pluginId } = this.currencyInfo
+    const keys = asSuiPrivateKeys(pluginId)(privateWalletInfo.keys)
+    return keys.mnemonic
+  }
+
+  async getDisplayPublicKey(publicWalletInfo: EdgeWalletInfo): Promise<string> {
+    const { keys } = asSafeCommonWalletInfo(publicWalletInfo)
+    return keys.publicKey
+  }
+
+  async importPrivateKey(input: string): Promise<JsonObject> {
+    const isValid = validateMnemonic(input)
+    if (!isValid) throw new Error('Invalid mnemonic')
+
+    // test mnemonic
+    Ed25519Keypair.deriveKeypair(input)
+
+    return { [this.networkInfo.pluginMnemonicKeyName]: input }
```

Comment:
Could we do:

```ts
const wasKeys = uncleaner(asSuiPrivateKeys(pluginId))
return wasKeys({mnemonic:input })
```

And then you can delete `pluginMnemonicKeyName` from the `networkInfo`. This is literally the whole reason we define `asSuiPrivateKeys` using `asCodec`, to handle the round-tripping.